Update Storage Management S3 onboarding (#36322)
DataDog/documentation: content/en/infrastructure/storage_management/amazon_s3.md

## Setup

Set up Storage Management for Amazon S3 with one of the following methods:

> - **Cloudformation**: A guided in-product setup that configures the AWS integration, enables S3 inventory on the buckets you select, and optionally enables S3 access logs. Setup launches a CloudFormation stack to apply changes in your AWS account.

Reviewer: ```suggestion
- **CloudFormation**: A guided in-product setup that configures the AWS integration, enables S3 Inventory on the buckets you select, and optionally enables S3 access logs. A CloudFormation stack applies the changes to your AWS account.
```

- **Terraform**: Use the official Datadog Storage Management Terraform module to configure inventory and access logs as code.
- **Manual**: Set up S3 inventory and the required permissions yourself in the AWS console, then register the inventory destination with Storage Management.

{{< tabs >}}

> {{% tab "Cloudformation" %}}

Reviewer: ```suggestion
{{% tab "CloudFormation" %}}
```

> In-product setup walks you through three steps: configuring an AWS account, selecting buckets and enabling S3 inventory and access logs, and finishing setup. A CloudFormation stack applies the required changes in your AWS account.

Reviewer: ```suggestion
The in-product setup walks you through three steps: configuring an AWS account, selecting buckets and enabling S3 Inventory and access logs, and finishing setup. A CloudFormation stack applies the required changes in your AWS account.
```

> To start, navigate to **Infrastructure** > [**Storage Management**][1] and click **Enable buckets**.

Reviewer: I don't see an **Enable buckets** option on the page opened by this link. Could you confirm if the UI has been updated?

[1]: https://app.datadoghq.com/storage-management

{{% collapse-content title="1. Configure AWS account" level="h4" expanded=false id="datadog-ui-step1" %}}

In this step, set up the Datadog AWS integration with metric and resource collection enabled.

1. Choose whether to use an **existing AWS account** already integrated with Datadog or to **add a new account**.
   - For a new account, a CloudFormation stack creates the Datadog integration role and configures both metric and resource collection.
   - For an existing account, confirm that **metric collection** and **resource collection** are enabled. Storage Management uses resource collection to discover S3 buckets and their existing inventory configurations.
2. Select the AWS region you want to configure. One region is configured per run; repeat the steps for each additional region.

> For a list of S3-related permissions used by resource collection, see [AWS Resource Collection][2].

Reviewer: ```suggestion
For a list of S3-related permissions used by resource collection, see [Resource collection][2] in the AWS integration page.
```

[2]: /integrations/amazon-web-services/#resource-collection
{{% /collapse-content %}}

{{% collapse-content title="2. Configure Storage Management" level="h4" expanded=false id="datadog-ui-step2" %}}

In this step, select the buckets to monitor, set an inventory destination, and optionally enable access logs.

<div class="alert alert-info">
    - Source bucket: The S3 bucket you want to monitor with Storage Management. <br>
    - Destination bucket: The bucket that stores inventory reports (one per AWS region, can be reused cross-account).
</div>

> 1. **Select buckets**: Choose the S3 buckets you want to monitor with Storage Management. Buckets already enabled for Storage Management are hidden. Buckets with existing S3 inventory are pre-selected and keep their current destination.

Reviewer: ```suggestion
1. **Select buckets**: Choose the S3 buckets you want to monitor with Storage Management. Buckets already enabled for Storage Management are hidden. Buckets with existing S3 Inventory are pre-selected and keep their current destination.
```

   - Select a destination bucket for access logs. You can use the same bucket as the inventory destination.
   - If a Datadog Log Forwarder is detected in the account, it is reused. Otherwise, the CloudFormation stack deploys a new forwarder.
   - Forwarded access logs can be ingested without indexing if used only for Storage Management. See [exclusion filters][3] for details.

   <div class="alert alert-warning">Forwarding S3 access logs to Datadog incurs Log Management ingestion costs. To minimize costs, use exclusion filters so logs are ingested but not indexed if used only for Storage Management. For details, see <a href="https://www.datadoghq.com/pricing/?product=log-management">Datadog Log Management pricing</a>.</div>

4. Click **Launch CloudFormation Template**. An AWS Quick Create stack opens, pre-filled with the bucket mappings, destination prefix, integration role name, and Datadog API key, application key, and log forwarder parameters.

5. In AWS, review the stack parameters and create the stack. The stack:

   - Enables daily S3 inventory on each selected source bucket.

>    - Adds IAM permissions for Storage Management to read the S3 inventory reports from the destination buckets.

Reviewer: ```suggestion
   - Enables daily S3 Inventory on each selected source bucket.
   - Adds IAM permissions for Storage Management to read the S3 Inventory reports from the destination buckets.
```

   - Adds the bucket policy to the inventory destination bucket so S3 can write inventory objects.
   - Enables S3 server access logging on selected buckets (if access logs are enabled).
   - Deploys a Datadog Log Forwarder Lambda function (if access logs are enabled and no forwarder exists).

[3]: /logs/log_configuration/indexes/#exclusion-filters
{{% /collapse-content %}}

{{% collapse-content title="3. Finish setup" level="h4" expanded=false id="datadog-ui-step3" %}}

> After the CloudFormation stack completes in AWS, return to Datadog and confirm setup.

Reviewer: What does a user actually do to confirm setup?

> The first inventory report can take up to 24 hours to generate. After that, your buckets appear in **Infrastructure** > [**Storage Management**][1].

Reviewer: The 24-hour wait is repeated in the **Validation** section (line `265`). Does it need to be mentioned here as well?

[1]: https://app.datadoghq.com/storage-management
{{% /collapse-content %}}

{{% /tab %}}

{{% tab "Terraform" %}}

> Use the official [Datadog Storage Management Terraform module][1] to configure S3 inventory and forward S3 access logs. The module configures the required permissions on the AWS integration IAM role, adds a bucket policy to allow Datadog to read inventory files from the destination bucket path, and enables S3 access log collection if a Datadog Log Forwarder is already set up.

Reviewer: ```suggestion
Use the official [Datadog Storage Management Terraform module][1] to configure S3 Inventory and forward S3 access logs. The module:

   - Configures the required permissions on the AWS integration IAM role.
   - Adds a bucket policy to allow Datadog to read inventory files from the destination bucket path.
   - Enables S3 access log collection if a Datadog Log Forwarder is already set up.
```

Reviewer: At 43 words, the second sentence here is well above the 25-word max guideline. Suggesting to break it up into a list to be more readable.

  # Bucket policy: "none", "create", or "merge" (default)
  destination_bucket_policy_management = "merge"

  # Optional: Enable S3 access logs for prefix-level request and latency metrics
  enable_access_logging           = true
  datadog_forwarder_function_name = "<DATADOG_FORWARDER_FUNCTION_NAME>"
}
```

> After enabling S3 inventory, it can take up to 24 hours for the first inventory reports to be generated. To check that inventories are being created, go to the AWS console, navigate to your destination bucket, and confirm that inventory files appear in the destination prefix you specified during setup.

Reviewer: ```suggestion
After enabling S3 Inventory, it can take up to 24 hours for the first inventory reports to be generated. To verify that inventories are being generated, go to your destination bucket in the AWS console and confirm that inventory files appear in the destination prefix you specified.
```

> After you confirm inventory files are present, verify Storage Management is enabled on your buckets by navigating to **Storage Management** > [**Enable Buckets**][2] and confirming that your destination bucket is listed and enabled.

Reviewer: The **Enable Buckets** link here doesn't seem to match the UI. Could you confirm if the link text could be updated to be more accurate?

[1]: https://registry.terraform.io/modules/DataDog/storage-management-datadog/aws/latest

> [2]: https://app.datadoghq.com/storage-monitoring?mConfigure=true&mStorageRecGroupBy=&mView=s3

Reviewer: ```suggestion
[2]: https://app.datadoghq.com/storage-management
```

Reviewer: Updating the link here. Not sure if the extra params are still needed; removing them didn't seem to have any impact on my test

[208]: /integrations/amazon-s3/#enable-s3-access-logs
{{% /collapse-content %}}

### Post-setup steps

After inventory files begin appearing in the destination bucket, enable Storage Management for that destination bucket by calling the following endpoint:

```bash
curl -X PUT "https://api.${DD_SITE}/api/v2/cloudinventoryservice/syncconfigs" \
  -H "Accept: application/vnd.api+json" \
  -H "Content-Type: application/vnd.api+json" \
  -H "DD-API-KEY: ${DD_API_KEY}" \
  -H "DD-APPLICATION-KEY: ${DD_APP_KEY}" \
  -d '{
    "data": {
      "id": "aws",
      "type": "cloud_provider",
      "attributes": {
        "aws": {
          "aws_account_id": "123456789012",
          "destination_bucket_name": "my-inventory-bucket",
          "destination_bucket_region": "us-east-1",
          "destination_prefix": ""
        }
      }
    }
  }'
```

Replace the request body values with:
- `aws_account_id`: The 12-digit AWS account ID that owns the destination bucket.
- `destination_bucket_name`: The name of the destination bucket holding inventory reports.
- `destination_bucket_region`: The AWS region of the destination bucket.

> - `destination_prefix`: The prefix within the destination bucket where inventory files are written. Use an empty string if there is no prefix.

Reviewer: These placeholders use a different format than the ones in the Terraform tab (lines `95` - `100`). Could you please update the placeholders here (and in the request body) to be consistent with the Terraform example? (The Terraform example is the preferred format.)

To verify your setup:
1. Wait for the first inventory report to generate (up to 24 hours for daily inventories).
2. Navigate to **Infrastructure** > [**Storage Management**][3] to see if the buckets you configured appear in the explorer list when **Monitored buckets** is selected.

  {{< img src="infrastructure/storage_management/monitored-buckets.png" alt="Validate bucket is enabled for monitoring" responsive="true">}}

### Best practices

Follow these best practices to optimize Storage Management setup:

> - **Configure lifecycle policies for inventory destination buckets**: S3 inventory reports are generated daily and stored in your destination bucket. To prevent old inventory files from accumulating and incurring storage costs, add a lifecycle policy to automatically delete inventory reports older than three days.

Reviewer: ```suggestion
- **Configure life cycle policies for inventory destination buckets**: S3 Inventory reports are generated daily and stored in your destination bucket. To prevent old inventory files from accumulating and incurring storage costs, add a life cycle policy to automatically delete inventory reports older than three days.
```

- **Configure lifecycle policies for S3 access logs**: If you have enabled S3 access logs for prefix-level request metrics, the raw log files accumulate in your destination bucket. After these logs are forwarded to Datadog, the raw files are no longer needed for Storage Management purposes. To automatically delete access log files after forwarding to Datadog, add a lifecycle rule.

  **Note**: Before enabling automatic deletion, verify that there are no compliance or audit requirements in your organization that mandate retaining raw S3 access logs for a specific period.

> - **Create exclusion filters for S3 access logs**: If S3 access logs are forwarded to Datadog only for Storage Management and don't need to be indexed for search or analytics, add an [exclusion filter][10] to keep them out of indexed log volume.

Reviewer: ```suggestion
- **Create exclusion filters for S3 access logs**: If S3 access logs are forwarded to Datadog only for Storage Management and don't need to be indexed for search or analytics, add an [exclusion filter][4] to keep them out of indexed log volume.
```

[10]: /logs/log_configuration/indexes/#exclusion-filters

> 

Reviewer: ```suggestion
```

Reviewer: Moving this to the bottom with other reference links, and renumbering to `[4]`

| aws.s3.inventory.access_logs.request_latency_by_method | `bucketname`, `prefix`, `region`, `method`                                                    | Server response time for requests in a prefix, optionally split by request method. Requires S3 access logs in Datadog.                          |

  *`prefixN` refers to prefix levels such as `prefix0`, `prefix1`, `prefix2`, and so on.

  **Note:** Use the right metric for the question you're answering:
  - `aws.s3.inventory.prefix_object_count` and `aws.s3.inventory.total_prefix_size` (with the `prefix` tag) include everything inside a folder and all its subfolders. Use these when you want the total count or size for a specific folder (for example, "how much is in `logs/2024/`?").
  - `aws.s3.inventory.prefix_object_count.levels` and `aws.s3.inventory.total_prefix_size.levels` (with `prefix1`, `prefix2`, `prefix3`, and so on) count or size objects only at that exact depth. Use these when you want to build a treemap or compare folder sizes across levels (for example, "which top-level folders are the biggest?").

>   **Note:** For the most accurate monitoring and visualization, use CSV inventory format and include all object versions to see non-current object recommendations or metrics.

Reviewer: ```suggestion
  **Note:** For the most accurate monitoring and visualization, include all object versions to see non-current object recommendations or metrics.
```

Reviewer: Line `56` states:
> Storage Management supports CSV inventory format only

However, this line seems to imply it's optional. Could you confirm and clarify this if needed?

### Prerequisites
Seeing recommendations requires the following prerequisites:
1. Configure S3 buckets for Storage Management by following the steps above on this page.
2. To see recommendations for moving infrequently accessed data to cheaper tiers by prefix, enable and forward S3 access logs to Datadog (Datadog Log Management fees apply).

> 3. To see recommendations for identifying non-current versions in prefixes, include "All versions" as part of the S3 inventory configuration.

Reviewer: ```suggestion
3. To see recommendations for identifying non-current versions in prefixes, include "All versions" as part of the S3 Inventory configuration.
```

  {{< img src="infrastructure/storage_management/storage-recs.png" alt="Storage Management Recommendations" responsive="true">}}

[1]: mailto:storage-monitoring@datadoghq.com

> [3]: https://app.datadoghq.com/storage-management

Reviewer: ```suggestion
[3]: https://app.datadoghq.com/storage-management
[4]: /logs/log_configuration/indexes/#exclusion-filters
```
